capsules/extra/src/hc_sr04.rs

```diff
@@ -0,0 +1,174 @@
+use kernel::hil::sensors::{self, Distance, DistanceClient};
+use kernel::hil::time::{Alarm, Ticks, Time};
+use kernel::hil::time::{AlarmClient, ConvertTicks, Frequency};
+use kernel::utilities::cells::OptionalCell;
+use kernel::ErrorCode;
+
+/// Maximum duration for the echo pulse to be measured in milliseconds.
+// As specified in the datasheet:
+// https://www.handsontec.com/dataspecs/HC-SR04-Ultrasonic.pdf,
+// the maximum time for the echo pulse to return is around 23 milliseconds
+// for a maximum distance of approximately 4 meters under standard temperature
+// and pressure conditions, but we use 38 milliseconds to account for variations
+// in real-world conditions. We use a slightly higher the value to account for
+// possible variations in measurement.
+pub const MAX_ECHO_DELAY_MS: u32 = 50;
+
+/// Speed of sound in air in mm/s.
+// The speed of sound is approximately 343 meters per second, which
+// translates to 343,000 millimeters per second. This value is used
+// to calculate the distance based on the time it takes for the echo
+// to return.
+pub const SPEED_OF_SOUND: u32 = 343000;
+
+/// Syscall driver number.
+use capsules_core::driver;
+pub const DRIVER_NUM: usize = driver::NUM::Distance as usize;
+
+#[derive(Copy, Clone, PartialEq)]
+/// Status of the sensor.
+pub enum Status {
+    Idle,         // Sensor is idle.
+    TriggerPulse, // Sending ultrasonic pulse.
+    EchoStart,    // Interrupt on the rising edge.
+    EchoEnd,      // Interrupt on the falling edge.
+}
+
+/// HC-SR04 Ultrasonic Distance Sensor Driver
+pub struct HcSr04<'a, A: Alarm<'a>> {
+    trig: &'a dyn kernel::hil::gpio::Pin,
+    echo: &'a dyn kernel::hil::gpio::InterruptPin<'a>,
+    alarm: &'a A,
+    start_time: Cell<u64>,
+    status: Cell<Status>,
+    distance_client: OptionalCell<&'a dyn sensors::DistanceClient>,
+}
+
+impl<'a, A: Alarm<'a>> HcSr04<'a, A> {
+    /// Create a new HC-SR04 driver.
+    pub fn new(
+        trig: &'a dyn kernel::hil::gpio::Pin,
+        echo: &'a dyn kernel::hil::gpio::InterruptPin<'a>,
+        alarm: &'a A,
+    ) -> HcSr04<'a, A> {
+        // Setup and return struct.
+        HcSr04 {
+            trig,
+            echo,
+            alarm,
+            start_time: Cell::new(0),
+            status: Cell::new(Status::Idle),
+            distance_client: OptionalCell::empty(),
+        }
+    }
+}
+
+impl<'a, A: Alarm<'a>> Distance<'a> for HcSr04<'a, A> {
+    /// Set the client for distance measurement results.
+    fn set_client(&self, distance_client: &'a dyn DistanceClient) {
+        self.distance_client.set(distance_client);
+    }
+    /// Start a distance measurement.
+    fn read_distance(&self) -> Result<(), ErrorCode> {
+        if self.status.get() == Status::Idle {
+            self.status.set(Status::TriggerPulse);
+            self.trig.set(); // Send the trigger pulse.
+            self.alarm
+                .set_alarm(self.alarm.now(), self.alarm.ticks_from_ms(1));
```

Comment:
There actually is a `ticks_from_us()` function that you could use to be more precise. I believe it rounds down though, so you might want `cmp::max(self.alarm.ticks_from_us(1), 1)`.

This is really just an optimization which will speed up measurements though, since 1 ms works fine now. So I don't think it's unreasonable to leave this as-is.

Either way, you should put a comment here documenting why "1" (or "10" if you change it) is a meaningful and correct choice.

Reply:
I've decided to keep it at 1 ms since the measurements are more accurate this way and I’ll add documentation to explain the reasoning.